File: 13-Roman-to-Integer.rs
impl Solution {
    /// Convert roman number to integer
    /// Time Complexity: O(n)
    /// Space Complexity: O(1)
    pub fn roman_to_int(s: String) -> i32 {
        // Will be stored calculated answer integer
        let mut result = 0;
        // For skipping next char when it finds special combination
        let mut skip_flg = false;

        // Scan a given roman number string
        for i in 0..s.len() {
            // Skip when there was special combination
            if skip_flg { skip_flg = false; continue; }

            // Check special combination
            let mut pair_number = s.get(i..i+2);
            if pair_number == Some("IV")      { result += 4;   skip_flg = true; }
            else if pair_number == Some("IX") { result += 9;   skip_flg = true; }
            else if pair_number == Some("XL") { result += 40;  skip_flg = true; }
            else if pair_number == Some("XC") { result += 90;  skip_flg = true; }
            else if pair_number == Some("CD") { result += 400; skip_flg = true; }
            else if pair_number == Some("CM") { result += 900; skip_flg = true; }
            else {
                // Convert 1 roman number char to integer and sum
                let curChar = s.get(i..i+1);
                if curChar == Some("I")      { result += 1;    }
                else if curChar == Some("V") { result += 5;    }
                else if curChar == Some("X") { result += 10;   }
                else if curChar == Some("L") { result += 50;   }
                else if curChar == Some("C") { result += 100;  }
                else if curChar == Some("D") { result += 500;  }
                else if curChar == Some("M") { result += 1000; }
            }
        }

        return result;
    }
}

// Sample Testcase:
//   Input:
//     "MCMXCIV"
//   Output:
//     1994

// Performance Result:
//   Runtime: 8 ms, faster than 100.00% of Rust online submissions for Roman to Integer.
//   Memory Usage: 2.4 MB, less than 87.50% of Rust online submissions for Roman to Integer.

File: 13-Roman-to-Integer/13-Roman-to-Integer.go
package main

import "fmt"

// Time Complexity : O(n)
// Space Complexity: O(1)
func romanToInt(s string) int {
    // Var for calculation result
    result := 0

    // Step value
    step := 1

    // Process for each chars
    for idx := 0; idx < len(s); idx += step {
        // Pick up max 2 length sub string
        pattern := fmt.Sprintf("%+v", s[idx:idx + 1])
        if idx + 2 <= len(s) {
            pattern = fmt.Sprintf("%+v", s[idx:idx + 2])
        }

        // Combination pattern check
        flgCombProcessed := false
        switch pattern {
            case "IV": { result += 4;   flgCombProcessed = true; }
            case "IX": { result += 9;   flgCombProcessed = true; }
            case "XL": { result += 40;  flgCombProcessed = true; }
            case "XC": { result += 90;  flgCombProcessed = true; }
            case "CD": { result += 400; flgCombProcessed = true; }
            case "CM": { result += 900; flgCombProcessed = true; }
        }

        // Single pattern check
        if !flgCombProcessed {
            switch pattern[0] {
                case 'I': result += 1
                case 'V': result += 5
                case 'X': result += 10
                case 'L': result += 50
                case 'C': result += 100
                case 'D': result += 500
                case 'M': result += 1000
            }
        }

        // Move 2 chars if this iteration's pattern is combination pattern
        if flgCombProcessed {
            step = 2
        } else {
            step = 1
        }

        // Debug Print
        // fmt.Printf("%d %+v %d\n", idx, pattern, result)
    }

    return result
}

// Sample Testcase:
//   Input:
//     MCMXCIV
//   Output:
//     1994

// Performance Result:
//   Runtime: 4 ms, faster than 91.09% of Go online submissions for Roman to Integer.
//   Memory Usage: 3.9 MB, less than 14.29% of Go online submissions for Roman to Integer.
